### DISCUSSION TITLE
AsyncServer _service_task was destroyed but it is pending when shutting down server

### DISCUSSION THREAD

**Discussion by sanderegg** [miguelgrinberg/python-socketio#1092] [General]
Hello,

I am using python-socketio together with aiohttp in a constellation of micro-services.

Lately I was checking some pytests that we have that are basically:
1. creating an aiohttp app with python-socketio AsyncServer attached to it.
2. connecting an AsyncClient to the server
3. and finishing the test
I see in pytest outputs these lines:
```bash
Task was destroyed but it is pending!
task: <Task pending name='Task-30' coro=<AsyncServer._service_task() running at /home/sanderegg/dev/github/osparc-simcore/.venv/lib/python3.9/site-packages/engineio/asyncio_server.py:498> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f553b04b2e0>()]>>
Task was destroyed but it is pending!
task: <Task pending name='Task-31' coro=<AsyncSocket.schedule_ping.<locals>.send_ping() running at /home/sanderegg/dev/github/osparc-simcore/.venv/lib/python3.9/site-packages/engineio/asyncio_socket.py:127> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f553b04b340>()]>>
```

The one with ```_service_task``` is systematic.
The one with ```schedule_ping``` is not but often there.

It is not making my test fail, but it does not look good.

while browsing into ```python-socketio``` and ```python-engineio``` I saw that these background tasks are actually never cancelled, or somehow I could not find anywhere how to properly close/delete/cleanup the AsyncServer. Is there such a possibility? Sorry if I missed it... but I could not manage to find anything regarding this.

The only way that I have found now, is to look though the current asyncio running tasks and find these to manually cancel them. but that is not very elegant. Is there any other option?

Thanks a lot for the very nice library!

**Comment by miguelgrinberg** [comment]
I have not implemented a graceful stop. To avoid the `service_task` warning you can create the server with `monitor_clients=False` when you are running tests. This will ensure this task is never started.

The `schedule_ping` task cannot currently be omitted, but if you reduce the `ping_interval` to say, one second, and then wait one second after all clients have been disconnected to let any pending tasks finish, then you should not see any warnings.

**Comment by sanderegg** [comment]
Thank you very much for the fast answer and for the tips. I guess I will continue with my manual cancelling then, this seems to work pretty well.
